---
marp: true
paginate: true
author: Vera Rykalina
theme: default
size: 4:3
footer: Vera Rykalina | July 2024

---

# HiVtype
## A Semi-Automated Workflow for HIV-1 Subtyping

---
## Contents

> Quick Guide

> Supplementary

> Conda

> GitHub Repo 

> Notes
---

# Quick Guide
---

# Important!

> Before you start the pipeline, make absolutely sure that all file names, sequence names, structures of input tables are correct and **follow exactly the patterns** as described in examples.

---

## Pipeline Directory
Locate to `Pipeline` directory:
```sh
$ cd ~/rki_subtyping/Pipeline
```
Open IDE (Visual Code Studio)
```sh
$ code ../
```
---
## Input Folders
Make sure you have 5 directories:

```sh
$ tree  -d
```

```sh
Pipeline/
├── AllSeqsCO20
├── InputFasta
├── Rega
├── Geno2Pheno
├── References
└── Scripts
```

---
## AllSeqsCO20 Folder
Provide this folder with .xlsx files as listed (from HiVpipe): 
```sh
AllSeqsCO20/
├── MS95_Seqs_ENV_CO20_V5.xlsx
├── MS95_Seqs_INT_CO20_V5.xlsx
└── MS95_Seqs_PRRT_CO20_V5.xlsx
```
Refer to Supplementary part of the guide, if you need more information (Page 21).

---
## InputFasta Folder
Provide this folder with files as listed (from HIVpipe): 
```sh
InputFasta/
├── MS95_ENV_20.fasta
├── MS95_INT_20.fasta
└── MS95_PRRT_20.fasta
```
Refer to Supplementary part of the guide, if you need more information (Page 22).

---



## Conda Environment
Activate `hivtype` environment.
```sh
$ conda activate hivtype
```
Be sure you have change in prompt:

```sh
(hivtype) ~/rki_sybtyping/Pipeline$
```
---
## Pipeline with **--outdir** parameter
Parameter `--outdir` determines a name of an ouput folder. The command will generate four enumerated output folders within `Results` folder. Without specifying an output folder you can get a warning message.
```sh
$ nextflow Scripts/hiv_type.nf --outdir Results
``` 

---

## Rega Folder (1)
Provide the folder with `.csv` files (separator: comma) generated by [Rega](https://www.genomedetective.com/app/typingtool/hiv) using marked `.fasta` files from the folder procuded by the pipeline:

`~/rki_subtyping/Pipeline/Results/1_marked_fasta` 

These `.fasta` files have M at the end of the file name:
```sh
1_marked_fasta/
├── MS95_ENV_20M.fasta
├── MS95_INT_20M.fasta
└── MS95_PRRT_20M.fasta
```
---

## Rega Folder (2)
Rega online tool always generates files with the same name **results**, e.g. `results.csv`. 
Rename these files accordingly, using the pattern as in the example below:

```sh
Rega/
├── Manual_Rega_MS95_ENV_20M.csv
├── Manual_Rega_MS95_INT_20M.csv
└── Manual_Rega_MS95_PRRT_20M.csv
```
Refer to Supplementary part of the guide, if you need more information (Page 23).

---
## Geno2Pheno Folder (1)
Provide the folder with `.csv` files (separator: comma) generated by [Geno2Pheno](https://subtyping.geno2pheno.org/) using marked `.fasta` files from the folder procuded by the pipeline:

`~/rki_subtyping/Pipeline/Results/1_marked_fasta` 

These `.fasta` files have M at the end of the file name:
```sh
1_marked_fasta/
├── MS95_ENV_20M.fasta
├── MS95_INT_20M.fasta
└── MS95_PRRT_20M.fasta
```
---

## Geno2Pheno Folder (2)
Geno2Pheno online tool always generates files with the same name **g2psubtypes_HIV**, e.g. `g2psubtypes_HIV.csv`. 
Rename these files accordingly, using the pattern as in the example below:

```sh
Geno2Pheno/
├── Manual_Geno2Pheno_MS95_ENV_20M.csv
├── Manual_Geno2Pheno_MS95_INT_20M.csv
└── Manual_Geno2Pheno_MS95_PRRT_20M.csv
```
---

## Pipeline with **--full** parameter
Repeat the previous command with `--full` parameter and 
`-resume` flag. The latter allows for generating an output up to `13_mafft` folder. The complete processes are cached.

```sh
$ nextflow Scripts/hiv_type.nf --outdir Results --full -resume
```

In the absence of the ENV-related files, also use the parameter **--noenv**
```sh
$ nextflow Scripts/hiv_type.nf --outdir Results --noenv --full -resume
```

Check the output of `13_mafft` folder before you run iqtree analysis (msa files - multiple sequence alignments)! 

---
## Pipeline with **--iqtree** parameter
Parameter `--iqtree` allows for running the iqtree process that produces `14_iqtree` folder within `Results`. The folder contains `.iqtree`, `.treefile`, and `.log` files. The parameter can be added at this point, as the last command with report output being produced or not added at all (no `14_iqtree` folder then).

```sh
$ nextflow Scripts/hiv_type.nf --outdir Results --full --iqtree -resume
``` 
You can monitor the `.log` file while running iqtree within `work` folder using respective process ID (68), e.g [68/72f0eb]. 

---

## Decision
Manually modify files (see below) which contain `Manual` tag in PRRT_Subpype, INT_Subtype, and ENV_Subtype columns. Save changes and close `.xlsx` files.

```sh
9_joint_with_tags/
├── full_MS95_ENV_20M.xlsx
├── full_MS95_INT_20M.xlsx
└── full_MS95_PRRT_20M.xlsx
```
---
## Report and Plot

```sh
$ nextflow Scripts/hiv_type.nf --outdir Results --full --iqtree -resume
``` 

Repeating the command above generates `15_report` folder with `MS95_uploads.xlsx` report file.

Repating it again generates a `MS95_subtype_counts.png` plot and adds it to the `15_report` folder.

---
## Clean Up

Once the pipeline has generated `Results` folder with all desired output files (**save** all needed outputs first), the input files can be removed from the input folders.
```sh
$ rm -rf InputFasta/* AllSeqsCO20/* Rega/* Geno2Pheno/* Results/
```
The same is true for dot and nextflow temp files/folders:
```sh
$ rm -rf .nextflow work .nextflow.log .nextflow.log.*
```

---
# Supplementary
---

## Example of .xlsx within AllSeqsCO20
```
Scount	        Fragment    Cutoff  Header	        Lauf   NGS-ID	Index GenBank-ID Sequenz
20-02944	PRRT	    20	    20-02944_PRRT_20	95	   	1	         CCCCT...
20-02945	PRRT	    20	    20-02945_PRRT_20	95	  	2	         CCCCT...
20-02947	PRRT	    20	    20-02947_PRRT_20	95	        3	         CCCCT...
20-02949	PRRT	    20	    20-02949_PRRT_20	95	        4	         CCCCT...
20-02950	PRRT	    20	    20-02950_PRRT_20	95	   	5	         CCCCT...
```
---

## Example of .fasta within InputFasta
```sh
>20-02955_ENV_20
GGAATTAGGCCAGTGGTGTCAACCCAACTATTGTTAAATGGCAGCCTAGCAGAAGAAGAT
GTGGTCATTAGATCTGAAAATTTCACAAACAATGCTAAAACCATAATAGTACAGCTTAAT
GAAACAGTAGTGATTAATTGTACAAGACCCGGCAACAATACAAGAAAAAGTATACATATA
GGACCAGGAAAAGCATGGTATGCAACAGGAGAGATAATAGGAGATATAAGACAAGCACAT
TGTAAACTTAATAAAACACAATGGGAAAAAACTTTAAAAAGGGTAGCTAGTAAATTAAGG
AAACAATCCAACCTTACAACAGTAATCTTTAAGAACTCCTCAGGGGGGGACCCAGAAATT
GTAATGCACAGTTTTAACTGTGGAGGGGAATTTTTCTATTGTAACACAACACAGTTGTTC
AATAGTATTTGGAATGACACTACTAATAGTACTGACACAAATGAAACTATCACACTCCCA
TGCAGAATAAAACAAATTATAAATAGATGGCAGGAAGCAGGAAGGG
```
---

## Example of .csv within Rega
An example of a `.csv` file produced by Rega online tool (names of columns and only one sample for demonstration)

```
"name","length","assignment","rule","support","begin","end","type","pure",
"pure_support","pure_inner","pure_outer","scan_best_support","scan_assigned_support",
"scan_assigned_nosupport","scan_best_profile","scan_assigned_profile","crf",
"crf_support","crf_inner","crf_outer","crfscan_best_support",
"crfscan_assigned_support","crfscan_assigned_nosupport","crfscan_best_profile",
"crfscan_assigned_profile","major_id","minor_id"
"20-02944_PRRT_20","1026.0","HIV-1 CRF 06_CPX","4","98.0","1823.0","2848.0","Human
immunodeficiency virus 1","HIV-1 Subtype G","93.0","0.0","93.0","0.5","0.357","0.643",
"G K A1 A1 A1 A1 G A1 A1 G G G G G","G - - - - - - - - G G G G -","HIV-1 CRF 06_CPX",
"98.0","0.0","98.0","1.0","1.0","0.0","06_CPX 06_CPX 06_CPX 06_CPX 06_CPX 06_CPX 
06_CPX 06_CPX 06_CPX 06_CPX 06_CPX 06_CPX 06_CPX 06_CPX","06_CPX 06_CPX 06_CPX 06_CPX 
06_CPX 06_CPX 06_CPX 06_CPX 06_CPX 06_CPX 06_CPX 06_CPX 06_CPX 06_CPX","",""
```
---

## Example of .csv within Geno2Pheno
An example of a `.csv` file produced by Rega online tool (names of columns and some samples for demonstration)

```
Sample,Result,Subtypes(Probability),Probability,Subtype (until break at position),virus
"24-01000_PRRT_20","Sequence is predicted (fit 97%) to be of HIV subtype A1.","A1(1)","0.97","A1(1026)","HIV"
"24-01001_PRRT_20","Sequence is predicted (fit 94%) to be of HIV subtype A2.","A2(1)","0.94","A2(1026)","HIV"
"24-01005_PRRT_20","Sequence is predicted (fit 99%) to be of HIV subtype A7.","A7(1)","0.99","A7(1026)","HIV"
"24-01006_PRRT_20","Sequence is predicted (fit 99%) to be of HIV subtype A8.","A8(1)","0.99","A8(1026)","HIV"
"24-01007_PRRT_20","Sequence is predicted (fit 99%) to be of HIV subtype B.","B(1)","0.99","B(1026)","HIV"
"24-01008_PRRT_20","Sequence is predicted (fit 96%) to be of HIV subtype C.","C(1)","0.96","C(1026)","HIV"
"24-01009_PRRT_20","Sequence is predicted (fit 98%) to be of HIV subtype D.","D(1)","0.98","D(1026)","HIV"
"24-01010_PRRT_20","Sequence is predicted (fit 96%) to be of HIV subtype F1.","F1(1)","0.96","F1(1026)","HIV"
"24-01017_PRRT_20","Sequence is predicted (fit 100%) to be of HIV subtype AE.","AE(1)","1","AE(1026)","HIV"
"24-01018_PRRT_20","Sequence is predicted (fit 93%) to be of HIV subtype G.","G(1)","0.93","G(1026)","HIV"

```
---

## References Folder
This folder contains reference panels and does not need any change unless reference panels should be replaced.   
```sh
References/
├── Reference_ENV_Panel.fas
├── Reference_INT_Panel.fas
└── Reference_PRRT_Panel.fas
```

---
# Conda

---
## Conda Info
List available conda environments.
```sh
$ conda info --envs
# conda environments:
#
base               * /opt/miniconda
hiv_type             /home/rykalinav/.conda/env/hivtype
```
---
## Conda Version
Pipeline's version of conda `23.3.1`
```sh
$ conda --version
```
---

## Deactivation of Environment
This command is used to deactivate the current invironment.
```sh
$ conda deactivate
```
Be sure you have change in prompt:

```sh
(base) ~/rki_sybtyping/Pipeline$
```

---
# GitHub Repo

---
## Repo Link 
The project is hosted [here](https://github.com/vera-rykalina/rki_subtyping.git).  Use this link to clone the repo in case of data loss.

---

## How to Clone
Locate to home directory
```sh
$ cd
```
Clone the repo 
```sh
$ git clone https://github.com/vera-rykalina/rki_subtyping.git
```
Modify path of `ProjectDir` within `hiv_type.nf`
Default option: projectDir = "$PWD"

---
# Notes
---
## Keep in Mind (1)

- for **--rkireport** only
The pipeline does not take into account subsubtypes. If there are subsubtypes they are converted to subtypes. For instance, A1 is converted to A, F2 is converted to F etc.

---
## Keep in Mind (2)
- The pipeline does not perform a full quality check of `.fasta` sequences. Illegal characters should be excluded (the pipeline takes care only of underscores so far). Sequences with illegal characters are not accepted by Rega online tool. 
---

## Keep in Mind (3)
- Make sure that sample names do not exceed 30 characters in length. Long sample names get shorten by Rega online tool that can cause issues. E.g, this sequence name **PK105_F482_23_MiS84_S86_20consensus_PRRT_20** is too long and gets shortern by Rega to **PK105_F482_23_MiS84_S86_20cons**. In such cases a manual change is necessary. 
---

## Keep in Mind (4)
- Be sure you are connected to the Internet

- You can always delete the whole `Results` folder or individual subfolder/subfolders within `Results` and repeat the commnand with `-resume`.
---

## Pipeline Updates
- Created: September 2022
- Updated: January 2023
A parameter **--noenv** has been added; should be used for the cases when ENV fragment is not sequenced. This info is reflected in the final report table.
- Updated: February 2023 
Slurm congif file added. Sierrapy client -> v.0.4.1
- Updated: March 2023
Sierrapy client -> v.0.4.2 (--no-sharding)
- Updated: June 2023
Sierrapy client -> v.0.4.3
